### ISSUE TITLE
worker_state is lost between map calls if the input is too large 

### ISSUE THREAD

**Issue by remondn-swk** [Slimmer-AI/mpire#46]
This is very related to #15.

Since your awesome release `v2.3.0` (which fix #15), I've been using `mpire` a lot, I love it :)

---

But I'm having a problem, very similar to #15.

In the following script, each worker get to deal with several numbers `i`, which they keep in state. Then I retrieve these values in another call.

```python
from mpire import WorkerPool


N = 12
W = 4


def set_state(w_state, i):
    w_state[i] = 2 * i + 1
    return None


def get_state(w_state, i):
    return w_state[i]


if __name__ == "__main__":
    pool = WorkerPool(n_jobs=W, use_worker_state=True, keep_alive=True)
    s = N // W

    pool.map(set_state, list(range(N)), iterable_len=N, n_splits=s)
    r = pool.map(get_state, list(range(N)), iterable_len=N, n_splits=s)

    print(r)
```

If I run this script, everything works perfectly, I get my expected output :

> [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23]

---

Now, if I change `N` (the number of tasks) to a higher number (like `128`) and I run the script again, I get the following error : 

> KeyError: '\n\nException occurred in Worker-2 with the following arguments:\nArg 0: 32\nTraceback (most recent call last):\n  File "/root/miniconda3/envs/housing_sb3/lib/python3.8/site-packages/mpire/worker.py", line 322, in _run_safely\n    results = func()\n  File "/root/miniconda3/envs/housing_sb3/lib/python3.8/site-packages/mpire/worker.py", line 276, in _func\n    results = func(args)\n  File "/root/miniconda3/envs/housing_sb3/lib/python3.8/site-packages/mpire/worker.py", line 415, in _helper_func_with_idx\n    return args[0], self._call_func(func, args[1])\n  File "/root/miniconda3/envs/housing_sb3/lib/python3.8/site-packages/mpire/worker.py", line 442, in _call_func\n    return func(args)\n  File "housing_drl/sb3/swag.py", line 14, in get_state\n    return w_state[i]\nKeyError: 32\n'

It's the exact same error as in #15, so it seems the worker state is somehow erased ?

---

@sybrenjansen Do you have any idea what's the problem ? Did I do something wrong in my script ?

**Comment by remondn-swk** [comment]
I just noticed that if I set the number of worker (`W`) to a higher number (say, `40`), it produces the right output.

I don't know what's going on behind the scenes, but this seems more like a dirty-workaround.

---

I'd like to be able to process my 128 tasks with a specific number of workers (like `4`) to not engorge my CPUs.

**Comment by remondn-swk** [comment]
Updated example, working perfectly with `multiprocessing.Barrier` :

```py
from mpire import WorkerPool
from multiprocessing import Barrier


N = 1024
W = 4
B = Barrier(W)


def set_state(w_state, i):
    w_state[i] = 2 * i + 1
    B.wait()
    return None


def get_state(w_state, i):
    B.wait()
    return w_state[i]


if __name__ == "__main__":
    pool = WorkerPool(n_jobs=W, use_worker_state=True, keep_alive=True)

    pool.map(set_state, list(range(N)), iterable_len=N, n_splits=W)
    r = pool.map(get_state, list(range(N)), iterable_len=N, n_splits=W)

    print(r)
```

Thanks again @sybrenjansen for the super clear explanation 🙏

**Comment by remondn-swk** [comment]
Actually, let me re-open this issue...

`Barrier` is working fine, but only if the number of tasks is "round".

If the number of tasks is not "round", then each worker will have to deal with a different number of tasks. In that case the code hangs (processes are waiting for W workers to be at the barrier, but only a subset are there).

**Comment by remondn-swk** [comment]
Ok, so I created a working example, by using 2 different barriers : One for the general case, and one for the last indexes (which might not be round).

I wrapped everything in a new class so it's easier to use : 

```py
from mpire import WorkerPool
from mpire.utils import chunk_tasks
from multiprocessing import Barrier


N = 1025
W = 4


class AdaptativeBarrier:
    def __init__(self, n_workers, n_tasks):
        self.general_barrier = Barrier(n_workers)
        self.last_barrier = Barrier(n_tasks % n_workers)

        # Compute the indexes of the last pass
        chunks = chunk_tasks(range(n_tasks), n_splits=n_workers)
        n = n_tasks // n_workers
        self.last_idx = [c[-1] for c in chunks if len(c) > n]

    def wait(self, i):
        if i in self.last_idx:
            return self.last_barrier.wait()
        else:
            return self.general_barrier.wait()


B = AdaptativeBarrier(W, N)


def set_state(w_state, i):
    w_state[i] = 2 * i + 1
    B.wait(i)
    return None


def get_state(w_state, i):
    B.wait(i)
    return w_state[i]


if __name__ == "__main__":
    pool = WorkerPool(n_jobs=W, use_worker_state=True, keep_alive=True)

    pool.map(set_state, list(range(N)), iterable_len=N, n_splits=W)
    r = pool.map(get_state, list(range(N)), iterable_len=N, n_splits=W)

    print(r)
```

---

I'll keep this issue open, as I think having an additional flag in `mpire` could be easier from a user perspective.

But feel free to close @sybrenjansen !

**Comment by sybrenjansen** [comment]
Let's keep it open. 

Like I said, it's not that much work and I do think it can be a useful feature for others as well. I think I can find some free time next week or so.

**Comment by sybrenjansen** [comment]
Had a bit less free time last few weeks than I thought. This is just a reminder that I didn't forgot about it, it's still on my to-do list

**Comment by sybrenjansen** [comment]
Now available in v2.5.0
